File: write-api-server-kotlin/src/main/kotlin/com/github/j5ik2o/adceet/api/write/adaptor/http/SwaggerDocService.kt
package com.github.j5ik2o.adceet.api.write.adaptor.http

import akka.http.javadsl.server.AllDirectives
import akka.http.javadsl.server.PathMatchers
import akka.http.javadsl.server.Route
import ch.megard.akka.http.cors.javadsl.CorsDirectives
import ch.megard.akka.http.cors.javadsl.settings.CorsSettings
import com.github.j5ik2o.adceet.api.write.adaptor.http.controller.ThreadController
import com.github.swagger.akka.javadsl.Converter
import com.github.swagger.akka.javadsl.SwaggerGenerator
import io.swagger.v3.oas.models.info.Info

class SwaggerDocService(private val host: String, private val port: Int) : AllDirectives() {

  private val generator = object : SwaggerGenerator {
    override fun converter(): Converter {
      return Converter(this)
    }

    override fun apiClasses(): MutableSet<Class<*>> {
      return mutableSetOf(
        ThreadController::class.java
      )
    }

    override fun apiDocsPath(): String {
      return "api-docs"
    }

    override fun host(): String {
      return "$host:$port"
    }

    override fun info(): Info {
      val info = Info()
      info.title = "thread service api"
      info.version = "v1"
      return info
    }

    override fun schemes(): MutableList<String> {
      return mutableListOf("http")
    }
  }

  fun toRoute(): Route {
    val route: Route = concat(
      path("swagger") {
        getFromResource("swagger/index.html")
      },
      getFromResourceDirectory("swagger"),
      path(
        PathMatchers.segment(generator.apiDocsPath()).slash("swagger.json")
      ) { get { complete(generator.generateSwaggerJson()) } },
    )

    val settings: CorsSettings = CorsSettings.defaultSettings()
    return CorsDirectives.cors(settings) { route }
  }
}

File: write-api-server-scala/src/main/scala/com/github/j5ik2o/adceet/api/write/http/SwaggerDocService.scala
package com.github.j5ik2o.adceet.api.write.http

import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.Route
import ch.megard.akka.http.cors.scaladsl.CorsDirectives
import ch.megard.akka.http.cors.scaladsl.settings.CorsSettings
import com.github.j5ik2o.adceet.api.write.http.controller.ThreadController
import com.github.swagger.akka.{ model, SwaggerGenerator }

class SwaggerDocService(private val hostName: String, private val port: Int) {

  val generator: SwaggerGenerator = new SwaggerGenerator() {

    override def apiClasses: Set[Class[_]] = Set(
      classOf[ThreadController]
    )

    override def schemes: List[String] = List("http")

    override def host: String = s"$hostName:$port"

    override def apiDocsPath: String = "api-docs"

    override def info: model.Info = {
      val info = model.Info()
      info.setTitle("thread service api")
      info.setVersion("v1")
      info
    }
  }

  def toRoute: Route = {
    val route: Route = concat(
      path("swagger") {
        getFromResource("swagger/index.html")
      },
      getFromResourceDirectory("swagger"),
      path(
        generator.apiDocsPath / "swagger.json"
      ) { get { complete(generator.generateSwaggerJson) } }
    )

    val settings: CorsSettings = CorsSettings.defaultSettings
    CorsDirectives.cors(settings) { route }
  }
}
